main/src/ca/uwaterloo/flix/language/phase/unification/BoolUnification.scala

```diff
@@ -59,10 +59,20 @@ object BoolUnification {
       case _ => // nop
     }
 
+    def cnf(t: Type): Type = {
+      import BoolMinimization._
+      val res = toType(toCNF(fromType(t)))
+      val before = t.toString.length
+      val after = res.toString.length
+      if (before > 9 || after > 9)
+        println(s"$before cnf'ed to $after")
+      res
+    }
+
     ///
     /// Run the expensive boolean unification algorithm.
     ///
-    booleanUnification(eraseAliases(tpe1), eraseAliases(tpe2))
+    booleanUnification(cnf(eraseAliases(tpe1)), cnf(eraseAliases(tpe2)))
```

Comment:
Rather, it is on the result you should call it. `booleanUnification` returns a potentially complicated subst. where we want to simplify each entry.

Reply:
I played a bit more. I get good results if I call `cnf` on l130 in BoolUnif. This line:

```
val st = Substitution.singleton(x.sym, cnf(mkOr(se(t0), mkAnd(x, mkNot(se(t1))))))
```